journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.instances.all._
+import cats.effect.implicits._
+import cats.effect.kernel.{Async, Deferred, Sync}
+import cats.effect.std.Mutex
+import cats.effect.{Ref, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+import java.time.Instant
+import java.util.UUID
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  type ObjectId = UUID
+  type WaiterId = UUID
+  type Waiter[F[_], O] = Deferred[F, Either[Throwable, (ObjectId, O)]]
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  )
+
+  def fixedSize[F[_]: Async, O](
+    resource: Resource[F, O],
```

Comment:
Not insisting, but I would consider:
1. Making it `=> Resource[F, O]` instead of `Resource[F, O]` as it will follow a least surprise principle and will allow one to add `println` into this block while debugging :grin: 
2. I would consider using two parameter lists, with `resource` being a last parameter to allow syntax like following:
```scala
ResourcePool.fixedSize(my,config,parameters) { creating resource here }
```

It is not a public API, so you are welcome to ignore this comment. I just wanted to share :blush:

Reply:
Thanks for the comments.
1. In what situations do you think `Resource[F, O]` instead of `=> Resource[F, O]` can be surprising?
2. Do you use `println` debugging? I don't use `println` even for local debugging in unit tests because then I have to think about how it plays together with cats-effect, so it's easier for me to do `log.debug` or `Sync[F].delay(println(...))`.